backend/app/services/ml_engine.py

```
@@ -0,0 +1,418 @@
+        "output_response": response,
+        "metadata": metadata,
+    }
+    try:
+        import fcntl
+        with TRACE_FILE.open("a", encoding="utf-8") as f:
+            fcntl.flock(f.fileno(), fcntl.LOCK_EX)
+            try:
+                f.write(json.dumps(trace, ensure_ascii=False) + "\n")
+            finally:
+                fcntl.flock(f.fileno(), fcntl.LOCK_UN)
+    except Exception as e:
```

Ruff (0.15.12)

[warning] Do not catch blind exception: `Exception`

(BLE001)